### ISSUE TITLE
Union type restricting to last assigned value in a hash

### ISSUE THREAD

**Issue by aktowns** [manastech/crystal#922]
Not sure if this is supposed to happen (and I can't have union typed generic classes in hashes)

```crystal
class Blah(T)
  def add(v: T)
  end
end

blahs = {} of Int32 => Blah(String | Int32)
blahs[0] = Blah(String).new 
blahs[1] = Blah(Int32).new

blahs[0].add("Hello")
blahs[1].add(0)
```

```
Error in line 10: no overload matches 'Blah(Int32)#add' with types String
Overloads are:
 - Blah(Int32)#add(v : Int32)
```

[example on play](http://play.crystal-lang.org/#/r/6at)

**Comment by bcardiff** [comment]
Well, there are some things going.

1) *You* know that the blah at 0 is of String. But the type inference can't distinguish between blah at 0 and blah at 1. They need to be have assigned the same static type (at least a union).

2) Actually I would expect that to compile. I wouldn't expect that `Foo(T) <: Foo(T | U)` for any generic class Foo. That is a `Foo(T)` can be used as a `Foo(T | U)`. @asterite is there any logic for this right now for the hashes values? Because http://play.crystal-lang.org/#/r/6ba shows a reduced example that does not work.

http://play.crystal-lang.org/#/r/6az declares the hash as `Int32 => Blah(String) | Blah(Int32)` which seems proper, but yet suffers from the same limitation detailed in 1).

What is the use case for a hash where the values might respond to one overload or another? There might be another way to code that?

**Comment by aktowns** [comment]
Thankyou for the response! 

My use-case is a bit convoluted, I'm working on an event system based on Xorg So I create a set of handlers using the Xorg event constants as keys [the actual code here](https://github.com/aktowns/rum/blob/master/src/rum/xorg/events.cr#L20). 

with the macros its a bit hard to follow but essentially unwraps to 
```crystal
@handlers[Constants::CreateNotify] = EventHandler(LibX11Events::XCreateWindowEvent).new(->(ev: Lib
X11Events::XEvent) { ev.xcreatewindow })
@handlers[Constants::UnmapNotify] = EventHandler(LibX11Events::XUnmapEvent).new(->(ev: LibX11Events::XEvent) { ev.xunmap })
```

The event object the Xlib uses is a C union and has a differing object based on the event fired. The initial plan was to register the set of handlers in the hash with a (generic)event class and a "retriever" block to pull the right object out of the union. (Not sure how much sense this makes!)

Essentially 
```crystal
WindowCreated = 0
KeyPressed = 1

@handler[WindowCreated] = MyEvent(NewWindow).new { |native_event| native_event.window} 
@handler[KeyPressed] = MyEvent(NewKeyPress).new ( |native_event| native_event.keypressed } 

@handler[WindowCreated].add -> (window: NewWindow) {
   puts "Woo a window!"
}

.. event loop .. 
  @handler[event.type].fire(event.native_event_union)
```

I guess I should rethink my approach :)

**Comment by bcardiff** [comment]
If you are dealing with external libs with it's own way to handle unions/dispatch you will most likely reach a point where a cast will be needed. It will be a bit of code that might explode in runtime instead of compile time.

As long as those casts do not reach the user every time I thinks is fine.

There might need to be hidden cast to handle the callback if you want to store them all together.

If all the events type are known, you might split them by types in different ivars and have a proper type.

Or by usage of macros you can wrap the `@handler[event.type]` in a way that returns casted versions of events based on the event type arg used at the handler registration.

(I'm just thinking out loud...)

**Comment by aktowns** [comment]
Ahh having separate ivars sounds like the 'correct' way just harder to manage, the macro/casted versions approach sounds a lot nicer to deal with!

thank you for your help

**Comment by asterite** [comment]
@bcardiff Maybe that shouldn't compile. Well, it doesn't compile, but you get an error in a different place. I tried to do a trivial thing to fix it but it seems it's not that simple.

@aktowns I think for this case, at least for now, you would need to store the handlers as a `Hash(Int32, {Void*, Void*})`, where `{Void*, Void*}` is the representation of a proc. Then you would define `on_...` handlers that accept blocks with restrictions (based on the event type). Then when invoking a handler you'd do `handler = @handlers[...]; Proc(ParticularEventObject).new(*handler)` to recreate the proc, but there you will have to have a big `case` for every event type and recreate the correct Proc type.

Eventually you'd be able to just say `Hash(Int32, Proc)` and that would mean the above, but you'd still have to do the casts.

(but this is just one solution I can think right now)

**Comment by jhass** [comment]
@aktowns so is the issue resolved for you? @asterite do you want to open a new issue about the thing that shouldn't compile?

**Comment by asterite** [comment]
I'll close this. Most of the issue is related to variance and covariance: #1297.
